lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -13,184 +13,217 @@
-            }
-            catch (CancellationException e) {
-                // Invalidated, or load failed
-            }
-            catch (ExecutionException e) {
-                // Should never happen
-                throw new IllegalStateException("Future unexpectedly completed with exception", e);
+            Map<Token<K>, V> values = dataCache.getAll(keyToToken.values());
+
+            BiMap<Token<K>, K> tokenToKey = keyToToken.inverse();
+            ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+            for (Map.Entry<Token<K>, V> entry : values.entrySet()) {
+                Token<K> token = entry.getKey();
+
+                // While token.getKey() returns equal key, a caller may expect us to maintain key identity, in case equal keys are still distinguishable.
+                K key = tokenToKey.get(token);
+                checkState(key != null, "No key found for %s in %s when loading %s", token, tokenToKey, keys);
+
+                result.put(key, entry.getValue());
             }
-            catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RuntimeException("Interrupted", e);
+            return result.buildOrThrow();
+        }
+        catch (Throwable e) {
+            for (Token<K> token : newTokens) {
+                // Failed to load and it was our new token persisted in tokens map.
+                // No cache entry exists for the token (unless concurrent load happened),
+                // so we need to remove it.
+                tokens.remove(token.getKey(), token);
             }
-
-            // Someone else was loading the key, but the load was invalidated.
+            throw e;
         }
     }
 
     @Override
-    public void put(K key, V value)
+    public long size()
     {
-        throw new UnsupportedOperationException("The operation is not supported, as in inherently races with cache invalidation. Use get(key, callable) instead.");
+        return dataCache.size();
     }
 
     @Override
-    public void invalidate(Object key)
+    public void cleanUp()
     {
-        delegate.invalidate(key);
+        dataCache.cleanUp();
     }
 
-    @Override
-    public void invalidateAll(Iterable<?> keys)
+    @VisibleForTesting
+    long tokensCount()
     {
-        delegate.invalidateAll(keys);
+        return tokens.size();
     }
 
     @Override
-    public void invalidateAll()
+    public void invalidate(Object key)
     {
-        delegate.invalidateAll();
+        @SuppressWarnings("SuspiciousMethodCalls") // Object passed to map as key K
+        Token<K> token = tokens.remove(key);
+        if (token != null) {
+            dataCache.invalidate(token);
+        }
     }
 
     @Override
-    public long size()
+    public void invalidateAll()
     {
-        // Includes entries being computed. Approximate, as allowed per method contract.
-        return delegate.size();
+        dataCache.invalidateAll();
```

Comment:
why did you choose to invalidate first the `dataCache` and not the `tokens` ?

Reply:
doesn't matter really

Reply:
Let's consider 2 threads: T1, T2

//invalidation call T1
Token<K> token = tokens.remove(key); // 1
if (token != null) {
     dataCache.invalidate(token);  // 3

// get call T2
Token<K> token = tokens.computeIfAbsent(key, ignored -> newToken); // 2
return dataCache.get(token); // 4

`tokens` map will contain an outdated useless token.

Per `com.google.common.cache.Cache` javadoc:
> @param <V> the type of the cache's values, which are not permitted to be null

In case that the value returned by `dataCache.get(token)` is `null` we should maybe remove remove the token from `tokens`. I'm not sure however if there's any gain in this.

Reply:
> `tokens` map will contain an outdated useless token.

that would be a problem, but i don't see this happening in your example.

Reply:
It does not look correct. It feels you can get leaked entry in  dataCache.
Consider following flow:
```
 * (T1) invalidateAll
 * (T1) dataCache.invalidateAll();
   *  entries from cache are gone
 * (T2) put(K, V)
   * (T2) load new value into `dataCache` for token which is present in the `tokens` map
 * (T1) `tokens.clear()`

Now we have an entry in `dataCache` for token which is no longer present in `tokens` map.

Note: maybe that is not possible if we `removalListeners` are executed synchronously as part of `dataCache.invalidateAll();`. But I am. not sure if that is true.

Reply:
Feel safer would be to get a copy of `tokens`; invalidate clear `tokens`; and then invalidate dataCache for each token one by one.

Reply:
The removalListener is called when evicting elements when the maximum size/weight on the cache has been reached and not on invalidation.
This can be easily checked with the tests provided in the PR

Reply:
> It does not look correct. It feels you can get leaked entry in dataCache.

correct, there is a small window where a value is loaded and subsequently made inaccessible.
this isn't a leak though, assuming the cache is bounded -- which was my assumption, but was not guaranteed.

I am adding a check in EvictableCacheBuilder mandating the cache to be bounded.

Reply:
> Feel safer would be to get a copy of `tokens`;

that would require switching away from `ConcurrentHashMap`, right?

Reply:
(T1) invalidateAll call
```
  dataCache.invalidateAll(); //1
  tokens.clear(); //4
```
T2 get(key, valueLoader) call
```
Token<K> token = tokens.computeIfAbsent(key, ignored -> newToken); //2
return dataCache.get(token, valueLoader); //3
```
After //4 the `dataCache` still contains an  "lost" entry for which there is no corresponding `tokens` entry.

Reply:
Reproduction scenario:

```

    @Test
    public void testInvalidateAllAndLoadConcurrently()
            throws Exception
    {
        AtomicLong remoteState = new AtomicLong(1);

        Cache<Integer, Long> cache = EvictableCacheBuilder.newBuilder().maximumSize(10_000).build();
        Integer key = 42;
        int threads = 1_0000;

        ExecutorService executor = newFixedThreadPool(threads);
        try {
            List<Future<Void>> futures = IntStream.range(0, threads)
                    .mapToObj(threadNumber -> executor.submit(() -> {
                        if (threadNumber%2 == 0) {
                            cache.get(key, remoteState::get);
                        }else{
                            cache.invalidateAll();
                            long cacheSize = cache.size();
                            if (cacheSize > 0){
                                System.out.println(cacheSize);
                            }
                        }

                        return (Void) null;
                    }))
                    .collect(toImmutableList());
            futures.forEach(MoreFutures::getFutureValue);

        }
        finally {
            executor.shutdownNow();
            executor.awaitTermination(10, SECONDS);
        }
    }
```

Reply:
> After //4 the `dataCache` still contains an "lost" entry for which there is no corresponding `tokens` entry.

absolutely.